In [ ]:
# nhập dữ liệu
import numpy as np
import pandas as pd
from scipy.stats import t
df = pd.read_csv("/content/drive/My Drive/Dataset/Lesson10_DataBTTH1.csv")
df.head()

In [ ]:
# định nghĩa output
y = df.values[:,3]
n=len(y)
y=y.reshape((n,1))
print(y)

In [54]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
# định nghĩa input
x = df.values[:,0:3]
x=x.reshape((n,3))
print(x)

In [57]:
# câu 1: uoc luong he so hoi quy
A=np.ones(len(y))
A=A.reshape((n,1))
X=np.concatenate((A,x),axis=1)
beta_hat = np.linalg.inv(X.T.dot(X)).dot(X.T).dot(y)
print(beta_hat)

[[-4.82511947]
 [-9.22716036]
 [ 0.63825804]
 [17.63203333]]


ước lượng củ hệ số hồi quy  $\beta = (\beta_0, \beta_1, \beta_2,\beta_3)$ là (-4.82511947; -9.22716036; 0.63825804; 17.63203333)

In [58]:
# câu 1: dự báo bien Y
Xnew=np.array([1, 14.5,225,5.5])
Xnew = Xnew.reshape(1,len(Xnew))
yhat1 = Xnew.dot(beta_hat)
print(yhat1)

[[101.9652982]]


dự báo hệ số khuếch đại của một thiết bị transistor HFE ($y$) với ba thông số: cực phát – RS $x_1 = 14.5$, cực gốc – RS $x_2 = 225$ và cực góp – RS $x_3 = 5.5$ là $\hat y =101.965298$

In [59]:
# câu 2: kiểm định mô hình HQTT
from scipy.stats import f
k=3
ybar = np.mean(y)
SST = sum((y-ybar)**2)
SSE = sum((y-yhat)**2)
SSR = SST - SSE
MSR = SSR/k
MSE = SSE/(n-k-1)
Fobs = MSR/MSE
print(Fobs)
p = 1-f.cdf(Fobs,k,n-k-1)
print(p)

[218.94524484]
[3.37729844e-13]


Vì p-value = 3.37729844e-13 << 0.05 nên ta kết luận rằng có mối quan hệ tuyến tính giữa biến phụ thuộc $y$ với các biến giải thích $x_1,x_2, x_3$.

In [60]:
# câu 3: kiểm định hệ số beta1
beta_hat1 = beta_hat[1]
sigma2 =(1/(n-k-1))*sum((y-yhat)**2)
C = np.linalg.inv(X.T.dot(X))
C11 = C[1,1]
Tobs = beta_hat1/np.sqrt(sigma2*C11)
print(Tobs)
p1 =2*(1-t.cdf(np.abs(Tobs),n-k-1))
print(p1)


[-1.2974294]
[0.21287926]


Vì p-giá trị = 0.21287926 >> 0.05 nên ta kết luận rằng hệ số khuếch đại của các thiết bị transistor HFE ($y$) không có quan hệ tuyến tính với thông số của cực phát – RS ($x_1$) 

In [61]:
# câu 4: he so xac dinh cua mo hinh
R2 = 1- SSE/SST
print(R2)
# he so xac dinh đa hieu chinh cua mo hinh
R2adj = 1- (SSE/(n-k-1))/(SST/(n-1))
print(R2adj)

[0.97622005]
[0.97176131]


Hệ số xác định mô hình $R^2 = 0.97176131$, mô hình HQTT có thể giải thích được 97% sự biến thiên của biến Y thông qua 3 biến $x_1,x_2, x_3$.